src/providers/pkcs11/capability_discovery.rs

```diff
@@ -0,0 +1,126 @@
+#![allow(unused, trivial_numeric_casts)]
+use super::Provider;
+use crate::authenticators::ApplicationName;
+use crate::providers::pkcs11::to_response_status;
+use cryptoki::types::mechanism::Mechanism;
+use cryptoki::types::mechanism::MechanismInfo;
+use cryptoki::types::mechanism::MechanismType;
+use cryptoki::types::Ulong;
+use log::trace;
+use parsec_interface::operations::can_do_crypto;
+use parsec_interface::operations::can_do_crypto::CheckType;
+use parsec_interface::operations::psa_algorithm::Algorithm;
+use parsec_interface::operations::psa_key_attributes::Attributes;
+use parsec_interface::operations::psa_key_attributes::EccFamily;
+use parsec_interface::operations::psa_key_attributes::Type;
+use parsec_interface::requests::ResponseStatus::{InvalidEncoding, PsaErrorNotSupported};
+use parsec_interface::requests::Result;
+use std::convert::TryFrom;
+use std::ops::Deref;
+
+impl Provider {
+    pub(super) fn can_do_crypto_internal(
+        &self,
+        _app_name: ApplicationName,
+        op: can_do_crypto::Operation,
+    ) -> Result<can_do_crypto::Result> {
+        let attributes = op.attributes;
+        let check_type = op.check_type;
+        let supported_ecc_family_sizes = [192, 224, 256, 384, 512];
+        match attributes.key_type {
+            Type::EccKeyPair {
+                curve_family: EccFamily::SecpR1,
+            }
+            | Type::EccPublicKey {
+                curve_family: EccFamily::SecpR1,
+            } => {
+                if !(supported_ecc_family_sizes.contains(&attributes.bits)) {
+                    return Err(PsaErrorNotSupported);
+                }
+            }
+            Type::RsaKeyPair | Type::RsaPublicKey => (),
+            _ => return Err(PsaErrorNotSupported),
+        }
+        match check_type {
+            CheckType::Generate => return Provider::generate_check(attributes),
+            CheckType::Import => return Provider::import_check(attributes),
+            CheckType::Use => {
+                if attributes.policy.permitted_algorithms == Algorithm::None {
+                    return Err(PsaErrorNotSupported);
+                } else {
+                    return Provider::use_check(&self, attributes);
+                }
+            }
+            CheckType::Derive => return Provider::derive_check(attributes),
+        };
+        return Ok(can_do_crypto::Result {});
+    }
+
+    fn use_check(&self, attributes: Attributes) -> Result<can_do_crypto::Result> {
+        if !(attributes.policy.usage_flags.decrypt()
+            || attributes.policy.usage_flags.encrypt()
+            || attributes.policy.usage_flags.sign_hash()
+            || attributes.policy.usage_flags.sign_message()
+            || attributes.policy.usage_flags.verify_hash()
+            || attributes.policy.usage_flags.verify_message())
+        {
+            return Err(PsaErrorNotSupported);
+        }
+        attributes.compatible_with_alg(attributes.policy.permitted_algorithms)?;
```

Comment:
Can you please make a PR on the book operation saying that `NotPermitted` is returned in that case if the alg is not compatible?